src/clj/kixi/hecuba/data/measurements/calculations.clj

```diff
@@ -434,3 +437,54 @@
+  [store calculated-sensor start end measurements]
+  (db/with-session [session (:hecuba-session store)]
+    (let [{:keys [sensor_id device_id]} calculated-sensor
+          calculated-data  [{:value (str (apply + (map :value (filter #(number? (:value %)) measurements))))
+                             :timestamp (tc/to-date end)
+                             :month (time/get-month-partition-key end)
+                             :device_id device_id
+                             :sensor_id sensor_id}]]
+      (measurements/insert-measurements store calculated-sensor 10 calculated-data)
+      (sensors/update-sensor-metadata session calculated-sensor start end))))
+
+(defn sum-up-expenditure-batch
+  "Fetch data for given range and underlying sensor type (calculated expenditure),
+  calculate total and insert it int othe database."
+  [store sensor start end calculation-type]
+  (db/with-session [session (:hecuba-session store)]
+    (let [{:keys [device_id sensor_id type]} sensor
+          new-type                           (sensors/output-type-for type calculation-type)
+          month                              (time/get-month-partition-key start)
+          where                              {:device_id device_id :sensor_id sensor_id :month month
+                                              :start start :end end}
+          measurements                       (measurements/parse-measurements (measurements/fetch-measurements store where))
+          output-sensor_id                   (:sensor_id (sensors/get-by-type {:device_id device_id :type new-type} session))]
+      (when (seq measurements)
+        (if output-sensor_id
+          ;; calculate total
+          (sum-up-expenditure store {:sensor_id output-sensor_id :device_id device_id} start end measurements)
+          (log/errorf "Could not find the output sensor_id for device_id %s and new type %s" device_id type))))))
+
+(defmulti total-usage (fn [_ item] (:usage-period item)))
+
+(defmethod total-usage :weekly
+  [store {:keys [sensor range]}]
+  (db/with-session [session (:hecuba-session store)]
+    (let [{:keys [start-date end-date]} range
+          sensor                        (sensors/get-by-id sensor session)]
+      (log/infof "Calculating total weekly usage for device_id %s and sensor_id %s" (:device_id sensor) (:sensor_id sensor))
+      (doseq [timestamp (time/seq-dates start-date end-date (t/days 7))]
+        (sum-up-expenditure-batch store sensor timestamp (t/plus timestamp (t/days 7)) :total-usage-weekly)))))
+
+(defmethod total-usage :monthly
+  [store {:keys [sensor range]}]
+  (db/with-session [session (:hecuba-session store)]
+    (let [{:keys [start-date end-date]} range
+          sensor                        (sensors/get-by-id sensor session)]
+      (log/infof "Calculating total monthly usage for device_id %s and sensor_id %s" (:device_id sensor) (:sensor_id sensor))
+      (doseq [timestamp (time/seq-dates start-date end-date (t/days 30))]
```

Comment:
after reading about the c* antipatterns and not using batches I am wondering if we should create a measurements-insert channel that we can just send messages to in an async way. I think we should do this as a separate piece of work though. wdyt?